File: MultiLanguageSandbox/src/test/deps/pascal/8.pas
{$ASSERTIONS ON}

uses
  SysUtils;
(*
Calculate the Hamming distance between two integers in binary representation.

Parameters:
- x (Integer): The first positive integer (x <= 1,000,000,000).
- y (Integer): The second positive integer (y <= 1,000,000,000).

Returns:
Integer: The Hamming distance between x and y, i.e., the number of bits that 
         need to be flipped to transform x into y in binary.
    >>> hamming_distance(1,2)
    2
*)

function hamming_distance(x, y: Integer): Integer;
var
  xor_result, count: Integer;
begin
  // Calculate the XOR of x and y
  xor_result := x xor y;
  
  // Count the number of set bits (1s) in the XOR result
  count := 0;
  while xor_result > 0 do
  begin
    if (xor_result and 1) = 1 then
      Inc(count);
    xor_result := xor_result shr 1;
  end;
  
  hamming_distance := count;
end;
procedure testHammingDistance;
begin
  Assert(hamming_distance(1, 2) = 2, 'Test Failed: (1, 2)'); // 01 and 10 have 2 different bits
  Assert(hamming_distance(4, 7) = 2, 'Test Failed: (4, 7)'); // 100 and 111 have 2 different bits
  Assert(hamming_distance(25, 30) = 3, 'Test Failed: (25, 30)'); // Additional test: 11001 and 11110 have 3 different bits
  Assert(hamming_distance(0, 0) = 0, 'Test Failed: (0, 0)'); // Additional test: Same numbers have 0 different bits
  Assert(hamming_distance($7FFF, 0) = 15, 'Test Failed: ($7FFF, 0)'); // Additional test: Max unsigned int and 0 have 32 different bits
end;

begin
  testHammingDistance;
end.

File: MultiLanguageSandbox/src/test/deps/typescript/8.ts

/**
 * Calculate the Hamming distance between two integers in binary representation.
 * Parameters:
 * - x (number): The first positive integer (x <= 1,000,000,000).
 * - y (number): The second positive integer (y <= 1,000,000,000).
 * Returns:
 * - number: The Hamming distance between x and y, i.e., the number of bits
 *           that need to be flipped to transform x into y in binary.
 * Example usage:
 *   hamming_distance(1, 2); // returns 2
 *   hamming_distance(4, 7); // returns 2
 */

function hamming_distance(x: number, y: number): number {
    let xor = x ^ y; // XOR operation to find differing bits
    let distance = 0;
    
    while (xor !== 0) {
        distance += xor & 1; // Check the least significant bit
        xor >>>= 1; // Right shift by 1 (unsigned)
    }
    
    return distance;
}
(() => {
    console.assert(hamming_distance(1, 2) === 2, '01 and 10 have 2 different bits');
    console.assert(hamming_distance(4, 7) === 2, '100 and 111 have 2 different bits');
    console.assert(hamming_distance(25, 30) === 3, '11001 and 11110 have 3 different bits');
    console.assert(hamming_distance(0, 0) === 0, 'Same numbers have 0 different bits');
    console.assert(hamming_distance(0xFFFFFFF, 0) === 28, 'Max unsigned int and 0 have 28 different bits');

    // If no assertion fails, log success message to the console
    // console.log("All tests passed!");
})();